src/sage/combinat/partition.py

```diff
@@ -6866,12 +6943,26 @@ def __iter__(self):
             sage: it = iter(P)
             sage: [next(it) for i in range(10)]
             [[], [1], [2], [1, 1], [3], [2, 1], [4], [3, 1], [2, 2], [5]]
+
+            sage: P = Partitions(inner=[2,2], outer=[3,2,1])
+            sage: list(P)
+            [[2, 2], [3, 2], [2, 2, 1], [3, 2, 1]]
         """
         n = 0
-        while True:
+        while n <= self._max_sum:
             for p in Partitions(n, **self._constraints):
                 yield self.element_class(self, p)
             n += 1
+    
+    def cardinality(self):
+        if self._max_sum == infinity:
+            return infinity
+        n = 0
+        total = 0
+        while n <= self._max_sum:
+            total += Partitions(n, **self._constraints).cardinality()
+            n += 1
+        return total
```

Comment:
```suggestion
        return sum(Partitions(n, **self._constraints).cardinality()
                   for n in range(self._max_sum+1))
```
Unfortunately, this creates a parent for every `n`.  I really think that we should get rid of `UniqueRepresentation` - not in this PR, however.

Also, this means that we go through `Partitions.__classcall_private__` for every `n`, but I hope that this is no bottleneck.

In any case, better cardinality functions could be implemented separately.  For example when only `outer` and `inner` are specified, the computation simplifies a lot - either using Kreweras' determinantal formula, or, more likely, using a recurrence.

Reply:
Yes, but already the current method goes though `__iter__` which is also creating a `Partititions` each time. So this would be potentially faster as instead of doing the `__iter__` of each `Partitions` object, if it has `cardinality` function, we can use that instead making it a little faster.

And as you mentioned, in future, we can do better cardinality things for specific constraints. This is just a starting point.